sat/cli/bootprep/input/image.py

```diff
@@ -479,3 +497,146 @@ def completed_successfully(self):
         return self.image_create_success and self.image_configure_success
 
     begin = begin_image_create
+
+
+class IMSInputImage(BaseInputImage, ABC):
+    """An input image that specifies an explicit IMS image or recipe as a starting point."""
+
+    @property
+    @abstractmethod
+    def ims_data(self):
+        """dict: the data under the 'ims' key"""
+
+    @property
+    def base_description(self):
+        """str: a human-readable description of the base we are starting from"""
+        for param in ('id', 'name'):
+            if param in self.ims_data:
+                return f'{self.base_resource_type} with {param}={self.ims_data[param]}'
+
+        # This shouldn't happen since schema requires 'id' or 'name'
+        return self.base_resource_type
+
+    @cached_property
+    def ims_base(self):
+        """dict: the data for the base IMS recipe or image to build and/or customize"""
+        resource_type = self.base_resource_type
+        try:
+            matching_base_resources = self.ims_client.get_matching_resources(
+                resource_type, resource_id=self.ims_data.get('id'),
+                name=self.ims_data.get('name'))
+        except APIError as err:
+            raise ImageCreateError(err)
+
+        if len(matching_base_resources) == 0:
+            raise ImageCreateError(f'Found no matches for {self.base_description}')
+        elif len(matching_base_resources) > 1:
+            raise ImageCreateError(f'Found {len(matching_base_resources)} matches for {self.base_description}. '
+                                   f'This {self.base_resource_type} must be specified by "id".')
+
+        return matching_base_resources[0]
+
+
+class IMSInputImageV1(IMSInputImage):
+    """An input image that specifies a base IMS image/recipe under the 'ims' property."""
+    @property
+    def ims_data(self):
+        """dict: the data that defines the base IMS image or recipe"""
+        return self.image_data['ims']
+
+    @property
+    def base_is_recipe(self):
+        """bool: whether the starting point is a recipe in IMS or an image"""
+        # is_recipe is a required property
+        return self.ims_data['is_recipe']
+
+
+class IMSInputImageV2(IMSInputImage):
+    """An input image that specifies a base IMS image/recipe under the 'ims' property under the 'base' property."""
+    @property
+    def ims_data(self):
+        """dict: the data that defines the base IMS image or recipe"""
+        return self.image_data['base']['ims']
+
+    @property
+    def base_is_recipe(self):
+        """bool: whether the starting point is a recipe in IMS or an image"""
+        # is_recipe is a required property
```

Comment:
nit: I think this is copy/paste cruft, should it instead be the following
```suggestion
        # type is a required property
```

Reply:
Yep, good catch!